lib/timer.py

```diff
@@ -0,0 +1,126 @@
+"""
+Timer script used to run all jobs that need to be frequently run on the system
+"""
+
+import json
+import logging
+import os
+
+import datetime
+
+from collections import OrderedDict
+
+from uaclient.config import UAConfig
+from uaclient import util
+from uaclient.jobs.update_messaging import update_apt_and_motd_messages
+
+try:
+    from typing import Any, Callable, Dict  # noqa: F401
+except ImportError:
+    # typing isn't available on trusty, so ignore its absence
+    def cast(_, x):  # type: ignore
+        return x
+
+
+LOG = logging.getLogger(__name__)
+
+# We are currently storing the time frequency for each job at this dict,
+# but we should move it later into uaclient.conf
+uaclient_jobs = OrderedDict(
+    {
+        "update_messaging": {
+            "job_func": update_apt_and_motd_messages,
+            "run_interval_seconds": 14400,
+        },
+        "update_status": {
+            "job_func": lambda cfg: None,
+            "run_interval_seconds": 14400,
+        },
+    }
+)
+
+
+def run_job(
+    job_name: str,
+    job_func: "Callable[[UAConfig], None]",
+    current_time: datetime.datetime,
+    run_interval_seconds: int,
+    jobs_status: "Dict[str, Any]",
+    cfg: UAConfig,
+) -> "Dict[str, Any]":
+    """Run a job based on time constraints.
+
+    If the current time is higher than the last time
+    the job has successfully executed plus the expected
+    time interval between job runs, we need to execute this
+    job again.
+
+    :param job_name: Name of the job to run.
+    :param job_func: Job function to run
+    :param run_interval_seconds: Interval in seconds that the job needs to
+                                 wait before running again.
+    :param jobs_status: A dict containing information about the last
+                        successful run of all jobs that should run
+                        in the system.
+
+    :return: A dict containing the update last successful run of the job
+             and the time it should execute next
+    """
+    seconds_to_add = datetime.timedelta(seconds=run_interval_seconds)
+    if job_name in jobs_status:
+        next_run = jobs_status[job_name]["next_run"]
+
+        if current_time < next_run:
+            return {}
+
+    try:
+        job_func(cfg)
+    except Exception as e:
+        LOG.warning("Error executing job: %s", job_name)
+        LOG.warning("The following exception was raised:\n%s", str(e))
```

Comment:
We could consolidate these two warnings into a single log
```suggestion
        LOG.warning("Error executing job: %s. %s", job_name, e)
```